src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5570,36 +5511,37 @@ private static Task InternalWhenAll(Task[] tasks)
         // which involves several allocations, with this logic:
         //      return new WhenAllPromise(tasksCopy);
         // which saves a couple of allocations and enables debugger notification specialization.
-        //
-        // Used in InternalWhenAll(Task[])
         private sealed class WhenAllPromise : Task, ITaskCompletionAction
         {
-            /// <summary>
-            /// Stores all of the constituent tasks.  Tasks clear themselves out of this
-            /// array as they complete, but only if they don't have their wait notification bit set.
-            /// </summary>
-            private readonly Task?[] m_tasks;
             /// <summary>The number of tasks remaining to complete.</summary>
             private int m_count;
+            /// <summary>The tasks that faulted.</summary>
+            private ConcurrentQueue<Task>? m_faultedTasks;
+            /// <summary>The first task that got canceled.</summary>
+            private Task? m_canceledTask;
 
-            internal WhenAllPromise(Task[] tasks)
+            internal WhenAllPromise(IEnumerable<Task?>? tasks)
             {
-                Debug.Assert(tasks != null, "Expected a non-null task array");
-                Debug.Assert(tasks.Length > 0, "Expected a non-zero length task array");
+                if (tasks == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.tasks);
 
                 if (TplEventSource.Log.IsEnabled())
                     TplEventSource.Log.TraceOperationBegin(this.Id, "Task.WhenAll", 0);
 
                 if (s_asyncDebuggingEnabled)
                     AddToActiveTasks(this);
 
-                m_tasks = tasks;
-                m_count = tasks.Length;
-
-                foreach (Task task in tasks)
+                int count = 0;
+                foreach (Task? task in tasks)
                 {
+                    if (task == null) ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
                     if (task.IsCompleted) this.Invoke(task); // short-circuit the completion action, if possible
                     else task.AddCompletionAction(this); // simple completion action
```

Comment:
What happens if the null input is discovered after we've already hooked up a continuation? Or if the enumerable itself throws? Doesn't that end up leaving a continuation registered to the previous tasks?

Reply:
Yes continuations would leak but only in cases where a null task is provided (which should be rare ?).

Reply:
> only in cases where a null task is provided

Or where enumeration fails with an exception (e.g. MoveNext throws), which is more likely. And if it happens, there's no way to recover / remove the anonymous continuation; that's my main concern, in that if for example the was some long-lived task involved, and used repeatedly, it could build up a long-list of dead continuations, which could also keep other state alive, such as another task involved in the operation.

Reply:
I couldn't find a use-case where we would reuse tasks after Task.WhenAll would throw (which would create that list of dead continuations). Because if you pass an IEnumerable to Task.WhenAll you expect to consume it so you won't reuse it. Did you have a precise case in mind?